src/xopen/__init__.py

```diff
@@ -330,21 +337,34 @@ def __iter__(self):
     def __next__(self) -> AnyStr:
         return self._file.__next__()
 
-    def _raise_if_error(self, allow_sigterm: bool = False) -> None:
+    def _raise_if_error(self, check_allowed_code_and_message: bool = False) -> None:
         """
-        Raise IOError if process is not running anymore and the exit code is
-        nonzero. If allow_sigterm is set and a SIGTERM exit code is
-        encountered, no error is raised.
+        Raise OSError if process is not running anymore and the exit code is
+        nonzero. If check_allowed_code_and_message is set, OSError is not raised when
+        (1) the exit value of the process is equal to the value of the allowed_exit_code
+        attribute or (2) the allowed_exit_message attribute is set and it matches with
+        the output from stderr.
         """
         retcode = self.process.poll()
-        if (
-            retcode is not None and retcode != 0
-            and not (allow_sigterm and retcode == -signal.SIGTERM)
-        ):
-            message = self._stderr.read().strip()
-            self._file.close()
-            self._stderr.close()
-            raise OSError("{} (exit code {})".format(message, retcode))
+        message = self._stderr.read().strip()
```

Comment:
I think I found the cause of the hang on macOS: The `read()` in this line blocks. Previously, it would only be called if the process has already terminated. In this version, it is called even if the process is still running. (`retcode` is `None`).

You need to move this line to above `if check_allowed_code_and_message`

Reply:
I also figured that caused it so I moved it. However, I stumbled into a different issue. I started a new review to handle that one.